Default to BOLT 4 tlv payload format onions

ViktorTigerstrom opened the pull request:
Closes #1393 

Removed the creation of Legacy onion format payloads (`msgs::OnionHopDataFormat::Legacy`) from the `onion_utils`.

Just let me know If there still is a scenario where we would like to create Legacy format payloads, and I'll update the PR!

tnull commented:
Hm, even though only very few nodes seem to support the legacy format and it might be removed from BOLT 4 soon, it is probably still be a bit early to remove legacy format support entirely?
 
IIUC, for defaulting to TLVs, it would probably suffice to ensure that `NodeFeatures` initialize with `supports_variable_length_onion` set in cases for which we have not received the features explicitly via `NodeAnnouncement`?

ViktorTigerstrom commented:
> IIUC, for defaulting to TLVs, it would probably suffice to ensure that `NodeFeatures` initialize with `supports_variable_length_onion` set in cases for which we have not received the features explicitly via `NodeAnnouncement`?

Oh, yes that's probably a more correct interpretation of "defaulting tlv onions". Thanks a lot for the feedback @tnull!

I'll push a new PR with that functionality implemented instead in the coming days. I'll also edit the name of this so that it more correctly corresponds to what it actually does, so that it can be merged once BOLT4 is updated.

TheBlueMatt commented:
Tagging this `blocked on spec` as we probably want to do #1414 for now and we can revisit this pending the BOLTs getting updated and other implementations doing the same.

dunxen commented:
https://github.com/lightning/bolts/pull/962#issuecomment-1235404960

Not sure if anyone has been tracking the legacy usage from our side, but looks like we can rebase and merge this 🔜😄

ViktorTigerstrom commented:
Nice, thanks for keeping track @dunxen! I've been out traveling, but will pick this and my other PRs up again come Monday.

ViktorTigerstrom commented:
Rebased on upstream.
Added one new fixup commit [66886d2](https://github.com/lightningdevkit/rust-lightning/pull/1413/commits/66886d2664eab72ab26d69071893a6f9a37ca948), that updates old tests to account for the drop of support for creating legacy onions. One could argue that the entire `test_default_to_assume_support_of_variable_length_onions` test should be removed as well. I think it makes sense to keep it until the BOLT PR is merged, and all legacy onion code is removed as well.

Also added a fixup commit [a525de8](https://github.com/lightningdevkit/rust-lightning/pull/1413/commits/a525de8ca8c5ae2829dcd24db197839757f77ea8), that updates the `test_always_create_onion_payload_tlv_format` test as well.

ViktorTigerstrom commented:
Thanks for the review @wpaulino!
Removed the `test_default_to_assume_support_of_variable_length_onions` test with [0705a97](https://github.com/lightningdevkit/rust-lightning/pull/1413/commits/0705a9782502b75e1b39670ca82de209736d671d)

I'm working on updating the test vectors as well. Yeah sure, I'll remove the `Legacy` enum variant in a follow-up PR!

dunxen commented:
Spec PR has been merged.

dunxen reviewed (commented):
Looks good :) Do we need a test where we do in fact receive a legacy payload and how we react to that?

ViktorTigerstrom commented:
Updated the test vectors with [23c6f72](https://github.com/lightningdevkit/rust-lightning/pull/1413/commits/23c6f72ad8b7123e2cff2a1c28ea0e3d56abd30e).

Let me know If you'd like me to push that commit in a separate PR to make this PR easier to review :).

ViktorTigerstrom commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -848,49 +843,95 @@ mod tests {
+
+		// Note that the resulting onion differs from the test vectors due to our
+		// `construct_onion_packet_with_init_noise` implementation.
+		assert_eq!(packet.encode(), hex::decode("0002EEC7245D6B7D2CCB30380BFBE2A3648CD7A942653F5AA340EDCEA1F283686619F7F3416A5AA36DC7EEB3EC6D421E9615471AB84C75A1B96D6BDE2F3BC5450EFB8B10E7EE7D90585AAF084E13A66E3B2BD0E926CCD8D263072206DBED0234F6505E21E282ABD8C0E4F5B9FF8042800BBAB065036EADD0149B37F27DDE664725A49866E052E809D2B0198AB9610FAA656BBF4EC516763A59F8F42C171B179166BA38958D4F51B3E73E6C7DD77BA8F874C6 ...
```
The result from packet.encode() differs from the test vectors, despite the `payloads`, `onion_keys` and `associated_data` are the equal to the test vector values.

I figured that this must be due to our `construct_onion_packet_with_init_noise` implementation, as our old implementation of the test vectors packet also differed from those vectors' onion. Is this correct?

I haven't had time to go through `construct_onion_packet_with_init_noise` deeply enough to understand why this is the case, so if anyone knows why, I'd be super thankful if you have time to explain why. Else I'll take the time to dig deeper into this later :)

ViktorTigerstrom replied:
Yeah agree, i'll debug it further and let you know the result. I'm a bit packed with work this week due to internal deadlines, but will try to squeeze it in as soon as I have time. Thanks for offering to look into it, but no worries, I'll do it as soon as I have time :)

ViktorTigerstrom reviewed (commented):
Thanks for the review @dunxen! I'll think through and look into that this weekend.

ViktorTigerstrom commented:
Dropped the `msgs::OnionHopDataFormat::Legacy`. I'm a little confused of how we want to handle backwards compatibility for legacy onions payloads that's already been serialized, so I would really appreciate some guidance on how handle that.

ViktorTigerstrom commented on lightning/src/ln/msgs.rs:
```diff
@@ -1489,56 +1478,51 @@ impl Readable for OnionHopData {
-				}
-			};
-			(format, amt.0, cltv_value.0)
+		if b.0 == LEGACY_ONION_HOP_FLAG {
```
How would should we handle this?

ViktorTigerstrom replied:
Nice thanks! [3778824](https://github.com/lightningdevkit/rust-lightning/pull/1413/commits/377882446afb3ed73339710336d30d2317c81dcf)

ViktorTigerstrom commented:
> > Dropped the msgs::OnionHopDataFormat::Legacy. I'm a little confused of how we want to handle backwards compatibility for legacy onions payloads that's already been serialized, so I would really appreciate some guidance on how handle that.
> 
> I dont think we need to care - we definitely dont expect users to have some pre-serialized onions lying around from a few years ago, and if they do, well, those onions aren't going to be useful as other nodes will start rejecting them anyway :)

Oh nice! And the fact that many implementations defaulted to the legacy format up until the beginning of this year should not impact this?

Addressed the latest feedback with [3778824](https://github.com/lightningdevkit/rust-lightning/pull/1413/commits/377882446afb3ed73339710336d30d2317c81dcf). I'll push merge conflict fixes separately soon now, to make it easy to differ the merge conflict fixes from the feedback fixes.

ViktorTigerstrom commented:
Latest push fixed merge conflicts

tnull commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -848,49 +843,95 @@ mod tests {
 		assert_eq!(onion_keys[4].mu, hex::decode("8e45e5c61c2b24cb6382444db6698727afb063adecd72aada233d4bf273d975a").unwrap()[..]);
 
-		// Test vectors below are flat-out wrong: they claim to set outgoing_cltv_value to non-0 :/
+		// Packet creation test vectors from BOLT 4 (as of writing this at bolt04/onion-test.json
```
Could you insert a Github permalink to the cited `onion-test.json` instead of using "as of writing"?

ViktorTigerstrom replied:
Thanks! Fixed with:
[ba522bf](https://github.com/lightningdevkit/rust-lightning/pull/1413/commits/ba522bfde17067d810481b04d8f6014836a101bf)

tnull reviewed (commented):
Looks good, one nit.

ViktorTigerstrom commented:
Oh nice, thanks for preparing that fuzz diff @TheBlueMatt!
Added it with [5a2ee11](https://github.com/lightningdevkit/rust-lightning/pull/1413/commits/5a2ee113b001837d9c6901a4578b10ac0ca62d9d). Let me know if you want it as a separate commit, or if it's ok as a fixup to the commit that removes the legacy format enum.

Also squashed all the previous fixups, I hope that was ok.

tnull commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -848,49 +843,95 @@ mod tests {
 		assert_eq!(onion_keys[4].mu, hex::decode("8e45e5c61c2b24cb6382444db6698727afb063adecd72aada233d4bf273d975a").unwrap()[..]);
 
-		// Test vectors below are flat-out wrong: they claim to set outgoing_cltv_value to non-0 :/
+		// Packet creation test vectors from BOLT 4 (as of writing
```
nit: Seems there is a leftover, and comment probably needs re-indentation.
```suggestion
		// Packet creation test vectors from BOLT 4 (see
```

ViktorTigerstrom replied:
Thanks! Do you like the new version better?

tnull replied:
Yeah, not the biggest fan of the overhanging line, but def. no blocker.

tnull reviewed (commented):
Tiny nit, looks good otherwise.  Feel free to squash fix-up commits.

ViktorTigerstrom commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -848,49 +843,94 @@ mod tests {
 		assert_eq!(onion_keys[4].mu, hex::decode("8e45e5c61c2b24cb6382444db6698727afb063adecd72aada233d4bf273d975a").unwrap()[..]);
 
-		// Test vectors below are flat-out wrong: they claim to set outgoing_cltv_value to non-0 :/
+		// Packet creation test vectors from BOLT 4 (see https://github.com/lightning/bolts/blob/16973e2b857e853308cafd59e42fa830d75b1642/bolt04/onion-test.json).
```
Ok @tnull, would you prefer the line like this instead? I'll change it if you think this is better :)

```suggestion
		// Packet creation test vectors from BOLT 4 (see 
		// https://github.com/lightning/bolts/blob/16973e2b857e853308cafd59e42fa830d75b1642/bolt04/onion-test.json).
```

ViktorTigerstrom replied:
This looks a bit off in the github GUI, but basically a new line right before the link.

tnull replied:
Yeah, I'd prefer moving it to the next line, but don't care too much to be honest.

ViktorTigerstrom replied:
Ok fixed :)

ViktorTigerstrom commented:
Thanks @TheBlueMatt! Agree, I'll look into that as soon as I have time.

ViktorTigerstrom commented:
Sorry for taking some time to get back and fixing this one! Added 77349f4 which makes the packet now match the test vectors :) 

> Hmm, so I assume its because the test vectors assume an initial noise value that is non-0, but we use a 0 value above?

You were correct :). The test vector initial payload bytes were deterministically derived corresponding to: https://github.com/lightningnetwork/lightning-rfc/pull/697, using the pad key-type (https://github.com/lightning/bolts/blob/b38156b9510c0562cf50f8758a64602cc0315c19/04-onion-routing.md#key-generation)

ViktorTigerstrom commented on lightning/src/ln/onion_utils.rs:
```diff
@@ -82,6 +82,15 @@ pub(super) fn gen_ammag_from_shared_secret(shared_secret: &[u8]) -> [u8; 32] {
 
+#[inline]
+#[allow(dead_code)]
+pub(super) fn gen_pad_from_shared_secret(shared_secret: &[u8]) -> [u8; 32] {
```
Would we like this to be test cfg?

ViktorTigerstrom commented:
Great, thanks! Changed the function to `cfg(test)` and squashed the fixup :)
